subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1222 @@
+ * @retval -EINVAL	If invalid buffer was provided or size is greater than already
+ *			allocated size.
+ */
+static int release_tx_blocks(struct backend_data *dev_data, size_t tx_block_index,
+			     size_t size, int new_size)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	struct block_header *block;
+	size_t num_blocks;
+	size_t total_size;
+	size_t new_total_size;
+	size_t new_num_blocks;
+	size_t release_index;
+	int r;
+
+	/* Calculate number of blocks. */
+	total_size = size + BLOCK_HEADER_SIZE;
+	num_blocks = (total_size + conf->tx.block_size - 1) / conf->tx.block_size;
+
+	if (new_size >= 0) {
+		/* Calculate and validate new values. */
+		new_total_size = new_size + BLOCK_HEADER_SIZE;
+		new_num_blocks = (new_total_size + conf->tx.block_size - 1) /
+				 conf->tx.block_size;
+		if (new_num_blocks > num_blocks) {
+			LOG_ERR("Requested %d blocks, allocated %d", new_num_blocks,
+				num_blocks);
+			return -EINVAL;
+		}
+		/* Update actual buffer size and number of block to release. */
+		block = block_from_index(&conf->tx, tx_block_index);
+		block->size = new_size;
+		release_index = tx_block_index + new_num_blocks;
+		num_blocks = num_blocks - new_num_blocks;
+	} else {
+		/* If size is negative, release all blocks. */
+		release_index = tx_block_index;
+	}
+
+	if (num_blocks > 0) {
+		/* Free bits in the bitmap */
+		r = sys_bitarray_free(conf->tx_usage_bitmap, num_blocks,
+				      release_index);
+		if (r < 0) {
+			LOG_ERR("Cannot free bits, err %d", r);
+			return r;
+		}
+
+		/* Wake up all waiting threads */
+		k_sem_give(&dev_data->block_wait_sem);
+	}
+
+	return tx_block_index;
+}
+
+/**
+ * Release all or part of the blocks occupied by the buffer.
+ *
+ * @param[in] buffer	Buffer to release
+ * @param[in] new_size	If less than zero, release all blocks, otherwise reduce size to
+ *			this value and update size in block header.
+ *
+ * @returns		Positive block index where the buffer starts or negative error
+ * @retval -EINVAL	If invalid buffer was provided or size is greater than already
+ *			allocated size.
+ */
+static int release_tx_buffer(struct backend_data *dev_data, const uint8_t *buffer,
+			     int new_size)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	size_t size;
+	int tx_block_index;
+
+	tx_block_index = buffer_to_index_validate(&conf->tx, buffer, &size);
+	if (tx_block_index < 0) {
+		return tx_block_index;
+	}
+
+	return release_tx_blocks(dev_data, tx_block_index, size, new_size);
+}
+
+/**
+ * Send data with ICMsg with mutex locked. Mutex must be locked because ICMsg may return
+ * error on concurrent invocations even when there is enough space in queue.
+ */
+static int icmsg_send_wrapper(struct backend_data *dev_data, uint8_t addr_or_msg_type,
+			      uint8_t tx_block_index)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	uint8_t message[2] = { addr_or_msg_type, tx_block_index };
+	int r;
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
```

Comment:
I don't see the point. Putting 2 bytes into ICMsg queue is not long or blocking operation, so we will not block here to often. Even if we block, we will serve higher priority thread fast, because of priority Inheritance on mutex. I don't see a real use case when something different than K_FOREVER makes sense.